File: MATTHEW/leetcode/AIcoding/2.Gemini-2.0/cpp_solutions/1593-split-a-string-into-the-max-number-of-unique-substrings.cpp
#include <string>
#include <unordered_set>

using namespace std;

class Solution {
public:
    int maxUniqueSplit(string s) {
        unordered_set<string> seen;
        return backtrack(s, 0, seen);
    }

private:
    int backtrack(string s, int start, unordered_set<string>& seen) {
        if (start == s.length()) {
            return 0;
        }

        int max_count = -1;
        for (int i = start; i < s.length(); ++i) {
            string sub = s.substr(start, i - start + 1);
            if (seen.find(sub) == seen.end()) {
                seen.insert(sub);
                int count = 1 + backtrack(s, i + 1, seen);
                max_count = max(max_count, count);
                seen.erase(sub);
            }
        }

        return max_count;
    }
};

File: leetcode/AIcoding/5.Qwen-3/python3_solutions/1593-split-a-string-into-the-max-number-of-unique-substrings.py
class Solution:
    def maxUniqueSplit(self, s: str) -> int:
        def backtrack(start, seen):
            if start == len(s):
                return len(seen)
            max_count = 0
            for end in range(start + 1, len(s) + 1):
                substr = s[start:end]
                if substr not in seen:
                    seen.add(substr)
                    max_count = max(max_count, backtrack(end, seen))
                    seen.remove(substr)
            return max_count

        return backtrack(0, set())